src/OrbitGl/ViewportTest.cpp

```diff
@@ -0,0 +1,176 @@
+
+  // Setting everything to the same values again should not mark the viewport as dirty
+  viewport.Resize(100, 200);
+  viewport.SetWorldWidth(500.f);
+  viewport.SetWorldHeight(600.f);
+  viewport.SetWorldTopLeftX(100.f);
+  viewport.SetWorldTopLeftY(-50.f);
+  viewport.SetWorldExtents(1000, 2000);
+  EXPECT_FALSE(viewport.IsDirty());
+}
+
+TEST(Viewport, ScrollingWithoutZoom) {
+  Viewport viewport(100, 200);
+  viewport.SetWorldExtents(200, 300);
+
+  viewport.SetWorldTopLeftX(100.f);
+  viewport.SetWorldTopLeftY(-50.f);
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -50.f));
+
+  viewport.SetWorldTopLeftX(150.f);
+  viewport.SetWorldTopLeftY(-150.f);
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -100.f));
+  EXPECT_EQ(viewport.IsDirty(), true);
+  viewport.ClearDirtyFlag();
+
+  // Scrolling further after clamping should not result in a dirty flag
+  viewport.SetWorldTopLeftX(200.f);
+  viewport.SetWorldTopLeftY(-200.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(100.f, -100.f));
+  EXPECT_EQ(viewport.IsDirty(), false);
+
+  // Resizing should clamp
+  viewport.SetWorldExtents(150, 250);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(50.f, -50.f));
+}
+
+TEST(Viewport, ScrollingWithZoom) {
+  Viewport viewport(100, 200);
+  viewport.SetWorldWidth(200);
+  viewport.SetWorldHeight(400);
+  viewport.SetWorldExtents(400, 800);
+
+  viewport.SetWorldTopLeftX(250.f);
+  viewport.SetWorldTopLeftY(-350.f);
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(200.f, -350.f));
+
+  viewport.SetWorldTopLeftX(300.f);
+  viewport.SetWorldTopLeftY(-450.f);
+
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(200.f, -400.f));
+  EXPECT_EQ(viewport.IsDirty(), true);
+  viewport.ClearDirtyFlag();
+
+  // Scrolling further after clamping should not result in a dirty flag
+  viewport.SetWorldTopLeftX(350.f);
+  viewport.SetWorldTopLeftY(-600.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(200.f, -400.f));
+  EXPECT_EQ(viewport.IsDirty(), false);
+
+  // Resizing should clamp
+  viewport.SetWorldExtents(250, 450);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(50.f, -50.f));
+}
+
+void VerifyConversion(Viewport& viewport, const Vec2i& screen_pos, const Vec2& world_pos,
+                      const Vec2& world_size) {
+  EXPECT_EQ(viewport.ScreenToWorldPos(screen_pos), world_pos);
+  EXPECT_EQ(viewport.ScreenToWorldHeight(screen_pos[1]), world_size[1]);
+  EXPECT_EQ(viewport.ScreenToWorldWidth(screen_pos[0]), world_size[0]);
+
+  EXPECT_EQ(viewport.WorldToScreenPos(world_pos), screen_pos);
+  EXPECT_EQ(viewport.WorldToScreenHeight(world_size[1]), screen_pos[1]);
+  EXPECT_EQ(viewport.WorldToScreenWidth(world_size[0]), screen_pos[0]);
+
+  EXPECT_EQ(viewport.QtToGlScreenPos(screen_pos),
+            Vec2i(screen_pos[0], viewport.GetHeight() - screen_pos[1]));
+}
+
+TEST(Viewport, CoordinateConversion) {
+  Viewport viewport(10, 100);
+  // Extents are large enough to scroll
+  viewport.SetWorldExtents(500, 500);
+  viewport.SetWorldWidth(20.f);
```

Comment:
A small comment here to explain the zoom-level?
Something like this:
// horizontal zoom_level = 2, vertical zoom-level = 0.5?

Reply:
Done